### PR TITLE
Install astropy to temporary directory for testing instead of building and copying

### PR DESCRIPTION
**Pull request by astrofrog**
This came out of a discussion on slack - basically now we are considering adding entry points, the issue is that entry points aren't installed when running the tests using ``python setup.py test``. After much searching/hacking, I realized that one way to solve this is that instead of building astropy and copying the build for testing, we can simply install it to a temporary directory, which should then take care of setting up the entry points. I'm pretty happy with this solution, it's no worse than what we had before and if anything a little cleaner since we aren't copying around build files.

@eteq @drdavella @saimn - what do you think?

@drdavella - can you check that this would fix things for the ASDF entry points?

Note that this doesn't help for the case where users do: ``pytest astropy``, but that's not currently an officially supported use case.

### CODE REVIEW

File: astropy/tests/command.py [hunk 1]
```diff
@@ -205,35 +205,43 @@ def run(self):
 
     def _build_temp_install(self):
         """
-        Build the package and copy the build to a temporary directory for
-        the purposes of testing this avoids creating pyc and __pycache__
-        directories inside the build directory
+        Install the package and to a temporary directory for the purposes of
+        testing this avoids creating pyc and __pycache__ directories inside the
+        build directory
```
Comment (saimn) [inline astropy/tests/command.py]: I think the pyc/pycache mention can be removed, since the goal is now to test also the `install` step.

**Comment by saimn** [comment]
As it creates the `.egg-info/` it should work for entry points, unless it needs to be added to `PYTHONPATH` ?

File: astropy/tests/command.py (hunk 1 above)
Comment (astrofrog) [reply astropy/tests/command.py]: Actually this is still relevant but I also expanded this to mention the other benefits.

**Comment by astrofrog** [comment]
@saimn - the entry points should work, I added a temporary check inside a test using pkg_resources and it looked like the entry points were there.

**Comment by astrofrog** [comment]
But @drdavella should confirm since he needs this for the ASDF entry point

**Comment by pllim** [comment]
This does not affect `astropy.test(...)`, correct?

**Comment by astrofrog** [comment]
@pllim - no, this doesn't change anything for ``astropy.test()`` when used from an installed version (which would therefore already have entry points anyway)

**Comment by astrofrog** [comment]
@eteq - thanks for the docs addition!

**Comment by eteq** [comment]
The gist of my "comment below is": it worries me a bit that now we have potentially different behavior for `astropy.test()` and `python setup.py test`. It'll be harder to try things with `astropy.test()` now because it will only work when installed... But I can live with that, as I imagine the vast majority of `astropy.test()` invocations are from an installed version anyway.

@drdavella - just to check on something related to the doc change I made: am I reading the situation right that your goal is basically to *replace* `python setup.py test <options>` with `pytest <options>`?  I.e., eventually we want to have pytest do all the stuff the test command currently does?  If so, FYI, you'll also need to update that doc section to no longer say "`python setup.py test` is the best way to deal with the install-and-build related subtleties".

**Comment by astrofrog** [comment]
> The gist of my "comment below is": it worries me a bit that now we have potentially different behavior for astropy.test() and python setup.py test. It'll be harder to try things with astropy.test() now because it will only work when installed... But I can live with that, as I imagine the vast majority of astropy.test() invocations are from an installed version anyway.

Just to be clear, this specific PR makes ``python setup.py test`` now be more in line with ``astropy.test()``, in that in both cases we now test an installed version, and in both cases entry points will be installed.

**Comment by eteq** [comment]
> this specific PR makes python setup.py test now be more in line with astropy.test()

Well, it depends: that's true *if* you call `astropy.test()` from an installed copy of astropy.  I was thinking about the following workflow (which I've used, but I think only when debugging the test runner):
```
python setup.py build
cd build/lib/whatever
python
>>> import astropy
>>> astropy.test()
```
but again, this isn't a deal-breaker, because I fully recognize this is a very edge-case.  Probably just something we (people who fiddle with the test machinery) need to be aware of, really.

**Comment by eteq** [comment]
All set, and I think we're all happy, so merging!

**Comment by bsipocz** [comment]
Can we backport this?

**Comment by Cadair** [comment]
A big 👍 to backporting this, would be really useful for SunPy and we will not be on 3.x for another release.

**Comment by eteq** [comment]
I have misgivings about backporting such a big change into the bugfix branch.  Nominally it's OK because in principal it's supposed to not be a change, but in practice I fear this will have ripple effects for affiliated packages, etc, that might be doing something not quite the same as astropy.

@Cadair - can you clarify that this *fixes* something in Sunpy? Or you think it's cleaner?

**Comment by eteq** [comment]
@astrofrog - what do you think about the backport question?

**Comment by Cadair** [comment]
Well I want to start using entry_points in sunpy before we will be able to rely on astropy 3.x. So if this doesn't get in to 2.0.x I will have to side-port it into sunpy.

**Comment by eteq** [comment]
Ah, gotcha.  I forgot that it's a necessity for using entry points.

The real question is if we’re *ever* going to pull it into the 2.0.x branch.  And I guess without doing this it’s impossible for anyone using 2.0.x to use entry_points.  And that's pretty much untenable so we might as well do it now.
